chain/network/src/peer_manager/tests/snapshot_hosts.rs

```diff
@@ -186,10 +186,25 @@ async fn invalid_signature_not_broadcast() {
         .send(PeerMessage::SyncSnapshotHosts(SyncSnapshotHosts { hosts: vec![info2.clone()] }))
         .await;
 
-    tracing::info!(target:"test", "Make sure that only the valid message is broadcast.");
+    tracing::info!(target:"test", "Make sure that only the valid messages are broadcast.");
+
+    loop {
+        let msg = peer2.events.recv_until(take_sync_snapshot_msg).await;
+        for info in msg.hosts {
+            if info == info2 {
+                return; // Received the expected message - end the test
+            }
+
+            // Even though `ok_info_a` and `ok_info_b` were sent in an invalid message,
+            // they were themselevs valid so the PeerManager can optionally accept them
+            // and broadcast them to other peers. This is an expected behavior, let's filter them out here.
+            if info == ok_info_a || info == ok_info_b {
+                continue;
+            }
 
-    let msg = peer2.events.recv_until(take_sync_snapshot_msg).await;
-    assert_eq!(msg.hosts, vec![info2]);
+            panic!("Unexpected host info received: {:#?}", info);
```

Comment:
It seems like we won't always detect this case because of the early exit above once the expected message `info2` is encountered.

These checks as a whole feel clunky and I wonder if you can do something nicer by making use of `assert_is_superset` as we do elsewhere to deal with the non-determinism from parallelization:

https://github.com/near/nearcore/blob/c6a4078b06dab1db4c4299161e1867908c313c60/chain/network/src/snapshot_hosts/tests.rs#L98-L101

IIUC you want to make sure that `hosts` contains `info2` and that it is a subset of `{ info2, ok_info_a, ok_info_b }`.

Reply:
The problem is that I'm waiting for multiple messages, and I don't know how many there will be, so I can't just collect them into a vec and call `assert_is_superset`.